scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala

```diff
@@ -0,0 +1,181 @@
+          "TDR",
+          "Born Digital",
+          "TDR",
+          List(
+            IdField("UpstreamSystemReference", tdrMetadata.FileReference),
+            IdField("BornDigitalRef", tdrMetadata.FileReference),
+            IdField("ConsignmentReference", tdrMetadata.ConsignmentReference),
+            IdField("RecordID", assetId.toString)
+          )
+        )
+        val file = FileMetadataObject(
+          fileId,
+          Option(assetId),
+          stripFileExtension(tdrMetadata.Filename),
+          1,
+          tdrMetadata.Filename,
+          tdrMetadata.ClientSideFileSize,
+          Preservation,
+          1,
+          fileLocation,
+          tdrMetadata.SHA256ServerSideChecksum
+        )
+        val metadataFileBytes = tdrMetadata.asJson.printWith(Printer.noSpaces).getBytes
+
+        val metadataChecksum = Stream
+          .emits(metadataFileBytes)
+          .through(fs2.hashing.Hashing[IO].hash(HashAlgorithm.SHA256))
+          .flatMap(hash => Stream.emits(hash.bytes.toList))
+          .through(fs2.text.hex.encode)
+          .compile
+          .to(string)
+
+        Stream.evals {
+          metadataChecksum.map { checksum =>
+            val metadataFileSize = metadataFileBytes.length
+            val metadata = FileMetadataObject(
+              metadataId,
+              Option(assetId),
+              s"${tdrMetadata.ConsignmentReference}-metadata",
+              2,
+              s"${tdrMetadata.ConsignmentReference}-metadata.json",
+              metadataFileSize,
+              Preservation,
+              1,
+              getMetadataUri(fileLocation),
+              checksum
+            )
+            List(archiveFolder, assetMetadata, file, metadata)
+          }
+        }
+      }
+    }
+
+    def downloadMetadataFile(lockTableMessage: LockTableMessage) = {
+      val fileLocation = lockTableMessage.location
+      val metadataUri = getMetadataUri(fileLocation)
+      dependencies.s3Client
+        .download(metadataUri.getHost, metadataUri.getPath.drop(1))
+        .map { pub =>
+          pub
+            .toStreamBuffered[IO](bufferSize)
+            .flatMap(bf => Stream.chunk(Chunk.byteBuffer(bf)))
+            .through(byteStreamParser[IO])
+            .through(fs2Decoder[IO, TDRMetadata])
+            .through(metadataStream => processTdrMetadata(metadataStream, fileLocation))
+        }
+    }
+
+    def processLockTableItems(lockTableRows: List[IngestLockTable]): IO[Unit] = {
+      Stream
+        .emits(lockTableRows)
+        .map(_.message)
+        .through(stringStreamParser[IO])
+        .through(fs2Decoder[IO, LockTableMessage])
+        .parEvalMap(config.concurrency)(downloadMetadataFile)
+        .parJoin(config.concurrency)
+        .compile
+        .toList
+        .flatMap { metadata =>
+          IO.raiseWhen(metadata.isEmpty)(new Exception(s"Metadata list for ${input.groupId} is empty")) >> {
```

Comment:
I don't think this is the place for that kind of validation. If it was to go anywhere it would go in your validation lambda but we'll see if we get any issues like that first.

Reply:
Ok. Yes, it is part of the validation lambda